### DISCUSSION TITLE
Postgres Exclude constraints

### DISCUSSION THREAD

**Discussion by AlexanderPodorov** [sqlalchemy/sqlalchemy#9924] [Usage Questions] [answered]
The example code from [here](https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#sqlalchemy.dialects.postgresql.ExcludeConstraint) does not work (I'm using Postgres 15):
```python
from sqlalchemy.dialects.postgresql import ExcludeConstraint, TSRANGE


class RoomBooking(Base):
    __tablename__ = "room_booking"

    room = Column(Integer(), primary_key=True)
    during = Column(TSRANGE())

    __table_args__ = (ExcludeConstraint(("room", "="), ("during", "&&")),)
```
The error thrown (at the moment when table is being created):
```
sqlalchemy.exc.ProgrammingError: (sqlalchemy.dialects.postgresql.asyncpg.ProgrammingError) <class 'asyncpg.exceptions.UndefinedObjectError'>: data type integer has no default operator class for access method "gist"
HINT:  You must specify an operator class for the index or define a default operator class for the data type.
[SQL: 
CREATE TABLE room_booking (
    room SERIAL NOT NULL, 
    during TSRANGE, 
    CONSTRAINT pk_room_booking PRIMARY KEY (room), 
    EXCLUDE USING gist (room WITH =, during WITH &&)
)

 ]
(Background on this error at: https://sqlalche.me/e/20/f405)
```
What could be missing there? Thanks!

**Comment by zzzeek** [comment]
I havent used this PG feature myself, you can maybe ask on the pg users list and show them that CREATE TABLE statement.   this error has to do with operator classes, some docs I found googling

https://www.postgresql.org/docs/current/sql-createtable.html#SQL-CREATETABLE-EXCLUDE
https://www.postgresql.org/docs/current/indexes-opclass.html

**Comment by lelit** [comment]
Strange, executing that statement directly with `psql` works as expected...

**Reply by zzzeek** [reply]
oh does it?  That's concerning then.    How about with an asyncpg connection directly ?

**Reply by zzzeek** [reply]
the error is described here: https://dba.stackexchange.com/questions/37351/postgresql-exclude-using-error-data-type-integer-has-no-default-operator-class

you get a different result with psql vs. SQLAlchemy ?

**Reply by lelit** [reply]
Sorry, can not try with SA, but with `psql` I get the following:
```
$ LANG= psql foo
psql (15.3 (Debian 15.3-1.pgdg+1))
Type "help" for help.

foo=# CREATE TABLE room_booking (
    room SERIAL NOT NULL, 
    during TSRANGE, 
    CONSTRAINT pk_room_booking PRIMARY KEY (room), 
    EXCLUDE USING gist (room WITH =, during WITH &&)
);
CREATE TABLE
```

The post uses PG 9, lots of moons ago...

**Reply by zzzeek** [reply]
OK well per the above linked SO answer there are [plugins](https://dba.stackexchange.com/a/37373) that need to be installed as well, so perhaps that's the issue here.

**Reply by lelit** [reply]
Indeed, `btree_gist` is automatically enabled on my Debian, at least I don't remember an explicit `create extension` on my part.

**Reply by AlexanderPodorov** [reply]
Just tried to execute `CREATE TABLE` directly and it works. I have `btree_gist` installed.
Slightly modified the original DDL (added `id` as promary key)
```sql
CREATE TABLE IF NOT EXISTS public.room_booking
(
    id integer NOT NULL DEFAULT nextval('room_booking_id_seq'::regclass),
    room integer,
    during tsrange,
    CONSTRAINT pk_room_booking PRIMARY KEY (id),
    CONSTRAINT room_booking_room_during_excl EXCLUDE USING gist (
        during WITH &&,
        room WITH =)

)
```
Exclude constraint works as expected:
![image](https://github.com/sqlalchemy/sqlalchemy/assets/54511144/bb71ce8d-4e6d-45e4-9a4e-aba67a41fa69)
Valid data can be inserted (different rooms can have overlapping datetime ranges):
![image](https://github.com/sqlalchemy/sqlalchemy/assets/54511144/868aa3c3-4a3d-4e57-abb3-1d83b01c2a53)
Since the error is:
```
sqlalchemy.exc.ProgrammingError: (sqlalchemy.dialects.postgresql.asyncpg.ProgrammingError) <class 'asyncpg.exceptions.UndefinedObjectError'>: data type integer has no default operator class for access method "gist"
```
Maybe it is an issue on `asyncpg` side?

**Reply by zzzeek** [reply]
> Slightly modified the original DDL (added id as promary key)

is that what makes it work?  we're looking for the **identical** DDL to work on psql as on asyncpg.   have we confirmed **identical** DDL works on psql and fails with raw asyncpg and/or SQLAlchemy asyncpg connection?  how about SQLAlchemy with psycopg and/or psycopg async as well?

**Reply by AlexanderPodorov** [reply]
Both ORM mappings fails on SQLAlchemy side. I adjusted the DDL in order to check actual exclude constraint (it was not possible with `room` being primary key). Related mapping is:
```python
from sqlalchemy.dialects.postgresql import ExcludeConstraint, TSRANGE


class RoomBooking(Base):
    __tablename__ = "room_booking"
    
    id = Column(Integer(), primary_key=True)
    room = Column(Integer())
    during = Column(TSRANGE())

    __table_args__ = (ExcludeConstraint(("room", "="), ("during", "&&")),)
```
I will try to test with `psycopg` later and report here.

**Reply by CaselIT** [reply]
I guess the original example does not make much sense, given that with a primary key on room it's impossible to trigger the exclude constraint.

Maybe it's this that the db is trying to communicate? (But it seems strange, since pg usually doesn't raise error like this unless there is a systax problem)

**Reply by AlexanderPodorov** [reply]
@CaselIT , yes you're right about the original example. However both original and modified examples raise the same error.
I updated the original example, since I wanted to make sure everything works as expected on my Postgres.

**Reply by CaselIT** [reply, marked as answer]
I've tried and once you enable the `btree_gist` the original snipped works as given

**Reply by CaselIT** [reply]
still it may make sense to improve that section

**Reply by AlexanderPodorov** [reply]
Oh, sorry, my bad, I forgot to add `btree_gist` on my test db (which I did on dev db).
Seems works as expected.
@CaselIT , yes, it makes perfect sense.
Thank you guys.

**Reply by CaselIT** [reply]
I've created https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4728 to mention that the extension is required
